test/Microsoft.Azure.SignalR.Tests/Infrastructure/ServiceConnectionProxy.cs

```diff
@@ -4,127 +4,169 @@
+using HandshakeResponseMessage = Microsoft.Azure.SignalR.Protocol.HandshakeResponseMessage;
 
 namespace Microsoft.Azure.SignalR.Tests
 {
-    internal class ServiceConnectionProxy
+    internal class ServiceConnectionProxy : IClientConnectionManager
     {
-        private static readonly IServiceProtocol ServiceProtocol = new ServiceProtocol();
-        private static readonly IInvocationBinder Binder = new InvocationBinder();
-        private static readonly HubMessage Message = new InvocationMessage("target", new object[] {"argument"});
-
-        private readonly IHubProtocol _hubProtocol;
-
-        public ConcurrentDictionary<string, int> ConnectionMessageCounter { get; } =
-            new ConcurrentDictionary<string, int>(StringComparer.Ordinal);
+        private static readonly TimeSpan DefaultHandshakeTimeout = TimeSpan.FromSeconds(5);
+        private static readonly IServiceProtocol _serviceProtocol = new ServiceProtocol();
 
         public IConnectionFactory ConnectionFactory { get; }
 
         public IClientConnectionManager ClientConnectionManager { get; }
 
-        public TestConnection ConnectionContext { get; }
+        private TestConnection ConnectionContext { get; }
 
         public ServiceConnection ServiceConnection { get; }
 
-        public ServiceConnectionProxy(string hubProtocolName = "json")
+        public ConcurrentDictionary<string, ServiceConnectionContext> ClientConnections => ClientConnectionManager.ClientConnections;
+
+        private readonly ConcurrentDictionary<string, TaskCompletionSource<ConnectionContext>> _waitForConnectionOpen = new ConcurrentDictionary<string, TaskCompletionSource<ConnectionContext>>();
+        private readonly ConcurrentDictionary<string, TaskCompletionSource<object>> _waitForConnectionClose = new ConcurrentDictionary<string, TaskCompletionSource<object>>();
+
+        public ServiceConnectionProxy()
         {
             ConnectionContext = new TestConnection();
             ConnectionFactory = new TestConnectionFactory(ConnectionContext);
             ClientConnectionManager = new ClientConnectionManager();
 
-            _hubProtocol = hubProtocolName.Equals("json", StringComparison.OrdinalIgnoreCase)
-                ? new JsonHubProtocol()
-                : (IHubProtocol) new MessagePackHubProtocol();
-
             ServiceConnection = new ServiceConnection(
-                ServiceProtocol,
-                ClientConnectionManager,
+                _serviceProtocol,
+                this,
                 ConnectionFactory,
                 NullLoggerFactory.Instance,
-                MessageCounterConnectionDelegate,
+                OnConnectionAsync,
                 Guid.NewGuid().ToString("N"));
         }
 
-        public void Start()
+        public Task StartAsync()
         {
             _ = ServiceConnection.StartAsync();
+            return HandshakeAsync();
         }
 
         public void Stop()
```

Comment:
Not saying it should change but it is kind of weird that Start is now async and Stop isn't.